### PR TITLE
Add Orth, AEC and N2D. Fix various bugs

### PR DESCRIPTION
**Pull request by collinleiber**
Add the alternative clustering algorithm Orth #9.

Introduce abstract Deep Clustering Class.
Add the alternative deep clustering algorithm AEC.
Add the sequential deep clusterin algorithm N2D.

Change the minimum version of Python to 3.10 (before 3.7).
Fix bug when setting the seed for Pytorch, where each run the same seed was used.
Fix bug in XMeans and GMeans, where number of clusters could exceed the maximum number of clusters #89.

### CODE REVIEW

File: clustpy/deep/_abstract_deep_clustering_algo.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+from clustpy.deep._utils import set_torch_seed
+from sklearn.base import BaseEstimator, ClusterMixin
+from sklearn.utils import check_random_state
+import numpy as np
+import torch
+
+
+class _AbstractDeepClusteringAlgo(BaseEstimator, ClusterMixin):
```
Comment (collinleiber) [inline clustpy/deep/_abstract_deep_clustering_algo.py:8]: Added this class which can be extended to included commonly used functionalities

File: clustpy/deep/_utils.py [hunk 2]
```diff
@@ -94,8 +94,29 @@ def detect_device(device: torch.device = None) -> torch.device:
     return device
 
 
-def encode_batchwise(dataloader: torch.utils.data.DataLoader, module: torch.nn.Module,
-                     device: torch.device) -> np.ndarray:
+def get_device_from_module(module: torch.nn.Module) -> torch.device:
```
Comment (collinleiber) [inline clustpy/deep/_utils.py:97]: Added this method to get the device of a torch model. Therefore, when encoding/decoding data we dont have to specify the device of the data anymore (must be the same as for the model)

File: clustpy/deep/_utils.py [hunk 3]
```diff
@@ -324,27 +338,31 @@ def run_initial_clustering(X: np.ndarray, n_clusters: int, clustering_class: Clu
...
-        else:
-            clustering_algo = clustering_class(n_components=n_clusters, **clustering_params)
-    else:  # in case of e.g., DBSCAN
-        if "random_state" in clustering_class_parameters and "random_state" not in clustering_params.keys():
-            clustering_algo = clustering_class(random_state=random_state, **clustering_params)
-        else:
-            clustering_algo = clustering_class(**clustering_params)
-    # Run algorithm
-    clustering_algo.fit(X)
+    if clustering_class is None:
```
Comment (collinleiber) [inline clustpy/deep/_utils.py:341]: clustering class can be None. In this case, random labels will be created

File: clustpy/deep/dcn.py [hunk 4]
```diff
@@ -278,22 +275,22 @@ def _loss(self, batch: list, autoencoder: torch.nn.Module, loss_fn: torch.nn.mod
             ae_loss, embedded, _ = autoencoder.loss(batch, loss_fn, device)
 
         # compute cluster loss
-        assignments = self.predict_hard(embedded)
-        assignment_matrix = int_to_one_hot(assignments, self.centers.shape[0])
-        cluster_loss = self.dcn_loss(embedded, assignment_matrix)
+        labels = self.labels[batch[0]]
```
Comment (collinleiber) [inline clustpy/deep/dcn.py:278]: We do not calculate the labels for the loss function but only when minibatch kmeans is executed. Corresponds to original implementation and the paper

File: clustpy/deep/dcn.py [hunk 5]
```diff
@@ -322,49 +321,40 @@ def fit(self, autoencoder: torch.nn.Module, trainloader: torch.utils.data.DataLo
...
                                   device)
                 # Backward pass - update weights
                 optimizer.zero_grad()
                 loss.backward()
                 optimizer.step()
-            # Update Assignments and Centroids
-            with torch.no_grad():
-                for batch in trainloader:
+                # Update Assignments and Centroids
+                with torch.no_grad():
```
Comment (collinleiber) [inline clustpy/deep/dcn.py:334]: we execute minibatch kmeans for each batch. Before we executed it after each epoch. Corresponds to original implementation

File: clustpy/deep/vade.py [hunk 6]
```diff
@@ -230,7 +237,7 @@ def __init__(self, n_clusters: int, embedding_size: int, weights: torch.Tensor =
             variances = torch.ones(n_clusters, embedding_size)
         assert variances.shape == (n_clusters,
                                    embedding_size), "Shape of the initial variances for the Vade_Module must be (n_clusters, embedding_size)"
-        self.p_var = torch.nn.Parameter(torch.tensor(variances), requires_grad=True)
+        self.p_log_var = torch.nn.Parameter(torch.log(torch.tensor(variances)), requires_grad=True)
```
Comment (collinleiber) [inline clustpy/deep/vade.py:240]: use log of initial variances

File: clustpy/deep/_utils.py [hunk 7]
```diff
@@ -20,7 +20,7 @@ def set_torch_seed(random_state: np.random.RandomState | int) -> None:
     if type(random_state) is int:
         seed = random_state
     elif type(random_state) is np.random.RandomState:
-        seed = (random_state.get_state()[1][0]).item()
+        seed = random_state.randint(np.iinfo(np.int32).max)
```
Comment (collinleiber) [inline clustpy/deep/_utils.py:23]: When using the old version, the random state will always be the same. Now it will be calculated randomly

File: clustpy/deep/_abstract_deep_clustering_algo.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        ----------
+        X: np.ndarray
+            The given data set
+
+        Returns
+        -------
+        X_embed : np.ndarray
+            The embedded data set
+        """
+        X_embed = self.autoencoder.transform(X, self.batch_size)
```
Comment (Lumik7) [inline clustpy/deep/_abstract_deep_clustering_algo.py:46]: Instead of the autoencoder here, should we already phrase it as generic encoder to  make this compatible with contrastive networks?

File: clustpy/deep/_train_utils.py [hunk 9]
```diff
@@ -81,15 +93,14 @@ def get_trained_autoencoder(trainloader: torch.utils.data.DataLoader, optimizer_
...
+                        optimizer_class=optimizer_class, loss_fn=loss_fn)
     if autoencoder.reusable:
         # If autoencoder is used by multiple deep clustering algorithms, create a deep copy of the object
         autoencoder = copy.deepcopy(autoencoder)
-
     return autoencoder
 
 
-def get_standard_initial_deep_clustering_setting(X: np.ndarray, n_clusters: int, batch_size: int,
+def get_standard_initial_deep_clustering_setting(X: np.ndarray | torch.Tensor, n_clusters: int, batch_size: int,
```
Comment (Lumik7) [inline clustpy/deep/_train_utils.py:103]: maybe rename to `get_default_deep_clustering_setting`?

File: clustpy/deep/_utils.py (hunk 2 above)
Comment (Lumik7) [reply clustpy/deep/_utils.py:97]: sounds goods. Does this fix the problem of clustpy of always choosing gpu:0 on multi-gpu systems as well? Because this was a problem I encountered quite often

File: clustpy/deep/vade.py (hunk 6 above)
Comment (Lumik7) [reply clustpy/deep/vade.py:240]: just be careful if you change this here you not to change the subsequent losses as well

**Review by Lumik7 — APPROVED** [review]
Only minor details need to be adapted. See comments

File: clustpy/deep/_utils.py (hunk 2 above)
Comment (collinleiber) [reply clustpy/deep/_utils.py]: Added an input parameter for choosing a specific GPU to all deep clustering algorithms

File: clustpy/deep/vade.py (hunk 6 above)
Comment (collinleiber) [reply clustpy/deep/vade.py:242]: All losses are calculated with exp(p_log_var). So the new version should be correct

File: clustpy/deep/_abstract_deep_clustering_algo.py (hunk 8 above)
Comment (collinleiber) [reply clustpy/deep/_abstract_deep_clustering_algo.py]: changed autoencoder to neural_network everywhere

File: clustpy/deep/_train_utils.py (hunk 9 above)
Comment (collinleiber) [reply clustpy/deep/_train_utils.py]: renamed to get_default_deep_clustering_initialization

File: clustpy/deep/dkm.py [hunk 10]
```diff
@@ -444,22 +458,20 @@ def __init__(self, n_clusters: int, alphas: tuple = (1000), batch_size: int = 25
...
+            "lr": 1e-4} if clustering_optimizer_params is None else clustering_optimizer_params
         self.pretrain_epochs = pretrain_epochs
         self.clustering_epochs = clustering_epochs
         self.optimizer_class = optimizer_class
         self.loss_fn = loss_fn
-        self.autoencoder = autoencoder
-        self.embedding_size = embedding_size
-        self.cluster_loss_weight = cluster_loss_weight
+        self.clustering_loss_weight = clustering_loss_weight
+        self.reconstruction_loss_weight = reconstruction_loss_weight
```
[the remark this answers is not in the document]
Comment (collinleiber) [reply clustpy/deep/dkm.py:475]: In my opinion, we should move this to a specific refactoring PR

File: clustpy/deep/tests/test_utils.py [hunk 11]
```diff
@@ -29,10 +29,26 @@ def test_squared_euclidean_distance():
 
 
 def test_detect_device():
-    # TODO idea for better test
+    # Test default
```
Comment (Lumik7) [inline clustpy/deep/tests/test_utils.py:32]: does this work in the CI pipeline?

**Review by Lumik7 — APPROVED** [review]
